src/eredis_cluster_manager.erl

```diff
@@ -0,0 +1,415 @@
+    version :: integer()  %% Used to avoid unnecessary refresh of Redis slots.
+}).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% API
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec start_link([]) -> {ok, pid()}.
+start_link({ModName, Args}) ->
+    gen_server:start_link({local, ModName}, ?MODULE, [Args], []).
+
+%% Looks up eredis Pid for Slot.
+-spec get_eredis_pid(Cluster::any(), Slot::integer()) -> 
+    {Pid::pid() | undefined, Version::integer()}.
+get_eredis_pid(Cluster, Slot) ->
+    {ok, Result} = gen_server:call(Cluster, {get_pid_by_slot, Slot}),
+    Result.
+
+
+%% Remaps various redis slots.
+-spec remap_cluster(Cluster::any(), Version::integer()) -> {Version::integer()}.
+remap_cluster(Cluster, Version) ->
+    {ok, Version} = gen_server:call(Cluster, {refresh_mapping, Version}),
+    Version.
+
+
+-spec qp(Cluster::any(), redis_pipeline_command()) -> redis_pipeline_result().
+qp(Cluster, Commands) ->
+    q(Cluster, Commands).
+
+
+-spec q(Cluster::any(), redis_command()) -> redis_result().
+q(Cluster, Command) ->
+    query(Cluster ,Command).
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Internal methods
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec get_pid_by_slot(State::#state{}, Slot::integer()) ->
+    {Pid::pid() | undefined, Version::integer()}.
+get_pid_by_slot(State, Slot) ->
+    Index = element(Slot + 1, State#state.slots),
+    Cluster = element(Index, State#state.slots_maps),
+    if
+        Cluster#slots_map.node =/= undefined ->
+            {ok, Pid} = lookup_eredis_pid(Cluster#slots_map.node#node.address,
+                                          Cluster#slots_map.node#node.port),
+            {Pid, State#state.version};
+        true ->
+            {undefined, State#state.version}
+    end.
+
+
+-spec refresh_mapping(State::#state{}, Version::integer()) -> State::#state{}.
+refresh_mapping(State, Version) ->
+    if State#state.version == Version ->
+        reload_slots_map(State);
+    true ->
+        State
+    end.
+
+
+%% TODO(vipin): Need to reset connection on Redis node removal.
+-spec reload_slots_map(State::#state{}) -> State::#state{}.
+reload_slots_map(State) ->
+    ClusterSlots = get_cluster_slots(State#state.init_nodes),
+    SlotsMaps = parse_cluster_slots(ClusterSlots),
+    ConnectedSlotsMaps = connect_all_slots(SlotsMaps),
+    Slots = create_slots_cache(ConnectedSlotsMaps),
+
+    NewState = State#state{
+        slots = list_to_tuple(Slots),
+        slots_maps = list_to_tuple(ConnectedSlotsMaps),
+        version = State#state.version + 1
+    },
+    NewState.
+
+-spec get_cluster_slots([#node{}]) -> [[bitstring() | [bitstring()]]].
+get_cluster_slots([]) ->
+    throw({error,cannot_connect_to_cluster});
+get_cluster_slots([Node|T]) ->
```

Comment:
when you are trying to get the cluster metadata I would recommend that we do it on a new connection just for this purpose. Trying to reuse the existing cluster connections might be an issue.

Reply:
I would like to reuse. I think reuse simplifies the code.

Reply:
Ok we can try this.